### PR TITLE
FlatGeobuf: Read feature indices on demand

### PR DESCRIPTION
**Pull request by bjornharrtell**
Aims to fix https://github.com/OSGeo/gdal/issues/1966

### CODE REVIEW

**Comment by bjornharrtell** [comment]
One test is failing:

=================================== FAILURES ===================================
____________________________ test_ogr_flatgeobuf_8 _____________________________

    def test_ogr_flatgeobuf_8():
    
        import test_cli_utilities
        if test_cli_utilities.get_test_ogrsf_path() is None:
            pytest.skip()
    
        ret = gdaltest.runexternal(test_cli_utilities.get_test_ogrsf_path() + ' -ro data/testfgb/poly.fgb')
    
>       assert ret.find('INFO') != -1 and ret.find('ERROR') == -1
E       AssertionError: assert (-1 != -1)
E        +  where -1 = <built-in method find of str object at 0x7fac3597a4b0>('INFO')
E        +    where <built-in method find of str object at 0x7fac3597a4b0> = '\nERROR ret code = -11'.find

ogr_flatgeobuf.py:279: AssertionError

File: gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp [hunk 1]
```diff
@@ -474,9 +471,14 @@ OGRFeature *OGRFlatGeobufLayer::GetFeature(GIntBig nFeatureId)
         m_ignoreSpatialFilter = true;
         m_ignoreAttributeFilter = true;
         uint64_t featureOffset;
-        auto err = readFeatureOffset(nFeatureId, featureOffset);
+        auto err = openFile();
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: it is a bit painful to attempt openFile() in so many places. I'd suggest rather to move that to dataset opening. Actually there's no need to try to open the file as it is already opened. It is sufficient to "steal" poOpenInfo->fpL: that is transfer it to the OGRFlatGeobufLayer object and setting poOpenInfo->fpL to nullptr afterwards.

Comment (bjornharrtell) [reply gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: Ah, I once tried to use the one from poOpenInfo but didn't know about the steal trick.. and the reason I tried to defer opening the file until needed is mostly due to lack of knowledge about the layer lifecycle (but I'm getting there). I'll try to refactor it in this PR, hang on.

**Comment by bjornharrtell** [comment]
Getting a new test failure :(

```
=================================== FAILURES ===================================
______________________________ test_ogr_basic_10 _______________________________

    def test_ogr_basic_10():
    
        import test_cli_utilities
        if test_cli_utilities.get_test_ogrsf_path() is None:
            pytest.skip()
    
        ret = gdaltest.runexternal(test_cli_utilities.get_test_ogrsf_path() + ' -all_drivers')
    
>       assert ret.find('INFO') != -1 and ret.find('ERROR') == -1
E       AssertionError: assert (0 != -1 and 4097 == -1
E        +  where 0 = <built-in method find of str object at 0x485ddb0>('INFO')
E        +    where <built-in method find of str object at 0x485ddb0> = 'INFO: TestCreate(PCIDSK).\nINFO: TestCreate(PDS4).\nINFO: TestCreate(VICAR).\nINFO: VICAR: Creation of /vsimem/test65...mem/test494 failed.\nINFO: Interlis 2: Creation of /vsimem/test495 failed.\nINFO: TestCreate(OGR\nERROR ret code = -11'.find
E         -4097
E         +-1)
```

**Comment by bjornharrtell** [comment]
It's `./test_ogrsf -all_drivers` that is segfaulting after `INFO: TestCreate(FlatGeobuf)`.

**Comment by bjornharrtell** [comment]
Strange, I would like to use gdb to find where the segfault occurs but `gdb --args ./test_ogrsf -all_drivers` just says `not in executable format: file format not recognized` even though I've configured and compiled with --enable-debug and I can use gdb with fx. ogrinfo.

**Comment by rouault** [comment]
> not in executable format: file format not recognized

might be the libtool wrapper if you didn't build --without-libtool. Then do `libtool --mode=execute gdb --args ./test_ogrsf -all_drivers`

**Comment by bjornharrtell** [comment]
That was it, thanks.

**Pull request merged**